src/categorical_algebra/CospanRewrite.jl

```diff
@@ -0,0 +1,241 @@
+
+const OpenGraphOb, OpenGraph = OpenCSetTypes(Graph, :V);
+
+"""
+Rewrite rule for structured cospans
+
+Left and right are intended to be discrete isomorphisms
+"""
+struct OpenRule
+    top::OpenGraph
+    middle::OpenGraph
+    bottom::OpenGraph
+    left::Span
+    center::Span # of Graph Transformations
+    right::Span
+
+    # Validate
+    function OpenRule(top::OpenGraph,
+                      middle::OpenGraph,
+                      bottom::OpenGraph,
+                      lft::Span,
+                      center::Span,
+                      rght::Span)
+
+        is_discrete = leg -> (nparts(dom(leg),:E)
+                            + nparts(codom(leg),:E) == 0)
+
+        is_iso = leg ->let f=collect(leg[:V]), n=length(f);
+                          (n == length(Set(f))
+                          && n == codom(leg[:V]).set) end
+
+        is_discrete_isospan = spn -> all(
+            leg -> is_iso(leg) && is_discrete(leg), spn)
+
+        @assert is_discrete_isospan(lft)
+        @assert is_discrete_isospan(rght)
+
+        for (quad, a,b,c,d) in [
+            ("upper left", left(lft),   left(top),     left(middle),  left(center)),
+            ("lower left", right(lft),  left(bottom),  left(middle),  right(center)),
+            ("upper right", left(rght),  right(top),    right(middle), left(center)),
+            ("lower right", right(rght), right(bottom), right(middle), right(center))]
+            @assert collect((a ⋅ b)[:V]) == collect((c ⋅ d)[:V]) "$quad $((a ⋅ b)[:V])!=$((c ⋅ d)[:V])"
+        end
+        return new(top, middle, bottom, lft, center, rght)
+    end
+end
+
+"""
+Based on figure at top of pg 15 of Cicala paper. Ignore "L" and
+replace primes with 0, 1, 2 (1 for center / no prime, 0 for
+top / prime, 2 for bottom / double prime)
+Underscores indicate morphisms
+
+Each OpenRule has 12 morphisms. The target starts out with only 2.
+"""
+function open_pushout_complement(
+    rule::OpenRule,
+    target::OpenGraph,
+    x0_y0::CSetTransformation, # match
+    a0_c0::CSetTransformation, # match
+    b0_d0::CSetTransformation  # match
+    )::OpenRule
+
+    # unpack input data
+    c0_y0, d0_y0 = left(target), right(target);
+    a0_x0, b0_x0 = left(rule.top), right(rule.top);
+    a1_x1, b1_x1 = left(rule.middle), right(rule.middle);
+    a2_x2, b2_x2 = left(rule.bottom), right(rule.bottom);
+    a1_a0, a1_a2 = left(rule.left), right(rule.left);
+    b1_b0, b1_b2 = left(rule.right), right(rule.right);
+    x1_x0, x1_x2 = left(rule.center), right(rule.center);
+
+    # compute results w/ DPO
+    @assert valid_dpo(b1_b0, b0_d0)
+    @assert valid_dpo(x1_x0, x0_y0)
+    @assert valid_dpo(a1_a0, a0_c0)
+    b1_d1, d1_d0 = pushout_complement(b1_b0, b0_d0);
+    x1_y1, y1_y0 = pushout_complement(x1_x0, x0_y0);
+    a1_c1, c1_c0 = pushout_complement(a1_a0, a0_c0);
+    d1_d2, b2_d2 = pushout(b1_d1, b1_b2);
+    y1_y2, x2_y2 = pushout(x1_y1, x1_x2);
+    c1_c2, a2_c2 = pushout(a1_c1, a1_a2);
```

Comment:
I tried a couple things but don't see a way to do that which clarifies (rather than obfuscates) what's going on. There are 9 pushouts and 3 pushout complements to compute, and it wasn't clear how to factor it into doing the same thing three times). I reorganized the code to highlight some common structure, as a first step.

Reply:
Yeah I think we just need to include the figure in the docs and put a reference to the figure in the docstring. There is now other way to unserstand this stuff, but to look at the diagrams.